src/main/java/graphql/execution/ExecutionStrategy.java

```diff
@@ -163,6 +168,76 @@ protected ExecutionStrategy(DataFetcherExceptionHandler dataFetcherExceptionHand
+
+        futures.await().whenComplete((completeValueInfos, throwable) -> {
+            BiConsumer<List<Object>, Throwable> handleResultsConsumer = handleFieldResults(executionContext, fieldNames, overallResult);
+            if (throwable != null) {
+                handleResultsConsumer.accept(null, throwable.getCause());
+                return;
+            }
+            Async.CombinedBuilder<Object> resultFutures = Async.ofExpectedSize(completeValueInfos.size());
+            for (FieldValueInfo completeValueInfo : completeValueInfos) {
+                resultFutures.add(completeValueInfo.getValue());
+            }
+            executionStrategyCtx.onFieldValuesInfo(completeValueInfos);
+            resultFutures.await().whenComplete(handleResultsConsumer);
+        }).exceptionally((ex) -> {
+            // if there are any issues with combining/handling the field results,
+            // complete the future at all costs and bubble up any thrown exception so
+            // the execution does not hang.
+            executionStrategyCtx.onFieldValuesException();
+            overallResult.completeExceptionally(ex);
+            return null;
+        });
+
+        instrumentationFuture.whenComplete(executionStrategyCtx::onCompleted);
+        return overallResult;
+    }
+
+    @NotNull
+    protected Async.CombinedBuilder<FieldValueInfo> resolveFieldsWithInfo(ExecutionContext executionContext, ExecutionStrategyParameters parameters, MergedSelectionSet fields, List<String> fieldNames) {
+        Async.CombinedBuilder<FieldValueInfo> futures = Async.ofExpectedSize(fields.size());
+        for (String fieldName : fieldNames) {
+            MergedField currentField = fields.getSubField(fieldName);
+
+            ResultPath fieldPath = parameters.getPath().segment(mkNameForPath(currentField));
+            ExecutionStrategyParameters newParameters = parameters
+                    .transform(builder -> builder.field(currentField).path(fieldPath).parent(parameters));
+
+            CompletableFuture<FieldValueInfo> future = resolveFieldWithInfo(executionContext, newParameters);
+            futures.add(future);
+        }
+        return futures;
+    }
+
+    private BiConsumer<List<Object>, Throwable> handleFieldResults(ExecutionContext executionContext, List<String> fieldNames, CompletableFuture<Object> overallResult) {
+        return (List<Object> results, Throwable exception) -> {
+            if (exception != null) {
+                handleInnerNonNullException(executionContext, overallResult, exception);
+                return;
+            }
+            Map<String, Object> resolvedValuesByField = new LinkedHashMap<>(fieldNames.size());
+            int ix = 0;
+            for (Object result : results) {
+                String fieldName = fieldNames.get(ix++);
+                resolvedValuesByField.put(fieldName, result);
+            }
+            overallResult.complete(resolvedValuesByField);
+        };
+    }
```

Comment:
This method is a copy of handleResults BUT the result signature is `CompletableFuture<Object> overallResult` and not a ER